In [1]:
import pandas as pd

In [2]:
# Dataset: http://guidetodatamining.com/chapter2/
df = pd.read_csv("Movie_Ratings.csv", encoding="cp1252")
df.head()

Unnamed: 0  Patrick C  Heather  Bryan  Patrick T  Thomas  aaron  vanessa  \
0         Alien        NaN      NaN    2.0        NaN     5.0    4.0      NaN   
1        Avatar        4.0      5.0    5.0        4.0     2.0    NaN      4.0   
2  Blade Runner        5.0      NaN    NaN        NaN     5.0    4.0      NaN   
3    Braveheart        4.0      NaN    5.0        NaN     4.0    4.0      3.0   
4     Dodgeball        5.0      4.0    3.0        2.0     4.0    NaN      4.0   

   greg  brian  ...   Zak  Matt  Chris.1  Josh  Amy  Valerie  Gary  Stephen  \
0   NaN    4.0  ...   NaN   NaN      4.0   3.0  NaN      NaN   2.0      5.0   
1   3.0    NaN  ...   5.0   NaN      NaN   4.0  3.0      2.0   1.0      4.0   
2   1.0    5.0  ...   NaN   NaN      3.0   NaN  3.0      3.0   1.0      NaN   
3   4.0    4.0  ...   5.0   NaN      4.0   NaN  3.0      4.0   5.0      5.0   
4   5.0    3.0  ...   3.0   NaN      3.0   NaN  4.0      3.0   4.0      3.0   

   Jessica  Jeff  
0      NaN   4.0  
1      NaN   4.0  
2      NaN   5.0  
3      NaN   4.0  
4      NaN   3.0  

[5 rows x 26 columns]

In [3]:
df = df.rename(columns={'Unnamed: 0': 'Movies'})
df.head()

Movies  Patrick C  Heather  Bryan  Patrick T  Thomas  aaron  vanessa  \
0         Alien        NaN      NaN    2.0        NaN     5.0    4.0      NaN   
1        Avatar        4.0      5.0    5.0        4.0     2.0    NaN      4.0   
2  Blade Runner        5.0      NaN    NaN        NaN     5.0    4.0      NaN   
3    Braveheart        4.0      NaN    5.0        NaN     4.0    4.0      3.0   
4     Dodgeball        5.0      4.0    3.0        2.0     4.0    NaN      4.0   

   greg  brian  ...   Zak  Matt  Chris.1  Josh  Amy  Valerie  Gary  Stephen  \
0   NaN    4.0  ...   NaN   NaN      4.0   3.0  NaN      NaN   2.0      5.0   
1   3.0    NaN  ...   5.0   NaN      NaN   4.0  3.0      2.0   1.0      4.0   
2   1.0    5.0  ...   NaN   NaN      3.0   NaN  3.0      3.0   1.0      NaN   
3   4.0    4.0  ...   5.0   NaN      4.0   NaN  3.0      4.0   5.0      5.0   
4   5.0    3.0  ...   3.0   NaN      3.0   NaN  4.0      3.0   4.0      3.0   

   Jessica  Jeff  
0      NaN   4.0  
1      NaN   4.0  
2      NaN   5.0  
3      NaN   4.0  
4      NaN   3.0  

[5 rows x 26 columns]

In [4]:
df['Movies'].nunique(), len(df.columns)

(25, 26)

In [5]:
data = df.copy()
data = data.set_index('Movies')
statistics_df = pd.DataFrame()
statistics_df['mean'] = data.mean(axis=1)

In [6]:
statistics_df['median'] = data.median(axis=1)

In [7]:
statistics_df['mode_first_term'] = data.mode(axis=1)[0]
statistics_df['mode_second_term'] = data.mode(axis=1)[1]

In [8]:
statistics_df['ratings_range'] = data.max(axis=1) - data.min(axis=1)

In [9]:
statistics_df['var'] = data.var(axis=1)

In [10]:
statistics_df['std'] = data.std(axis=1)

In [11]:
statistics_df['75th_quantile'] = data.T.quantile(0.75)

In [12]:
statistics_df['25th_quantile'] = data.T.quantile(0.25)

In [13]:
statistics_df['skewness'] = data.skew(axis=1)

In [14]:
statistics_df

mean  median  mode_first_term  mode_second_term  \
Movies                                                                      
Alien                 3.500000     4.0              4.0               NaN   
Avatar                3.550000     4.0              4.0               NaN   
Blade Runner          3.750000     4.5              5.0               NaN   
Braveheart            3.894737     4.0              4.0               NaN   
Dodgeball             3.619048     3.0              3.0               NaN   
Forest Gump           4.208333     4.0              5.0               NaN   
Gladiator             4.000000     4.0              4.0               NaN   
Jaws                  3.210526     3.0              3.0               NaN   
Kazaam                1.687500     1.0              1.0               NaN   
Lord of the Rings     3.454545     3.5              3.0               4.0   
Napolean Dynamite     2.708333     3.0              1.0               4.0   
Old School            4.000000     4.0              5.0               NaN   
Pootie Tang           1.777778     1.0              1.0               NaN   
Pulp Fiction          4.000000     4.0              4.0               NaN   
Scarface              3.666667     4.0              4.0               NaN   
Shawshank Redemption  4.500000     5.0              5.0               NaN   
Snakes on a Plane     2.133333     2.0              1.0               2.0   
Spiderman             3.500000     3.5              3.0               4.0   
Star Wars             4.608696     5.0              5.0               NaN   
The Dark Knight       4.421053     4.0              4.0               5.0   
The Happening         1.125000     1.0              1.0               NaN   
The Matrix            4.086957     4.0              4.0               NaN   
Toy Story             4.280000     4.0              4.0               NaN   
Village               2.533333     2.0              2.0               NaN   
You Got Mail          1.733333     2.0              2.0               NaN   

                      ratings_range       var       std  75th_quantile  \
Movies                                                                   
Alien                           3.0  1.388889  1.178511            4.0   
Avatar                          4.0  1.523684  1.234376            4.0   
Blade Runner                    4.0  2.386364  1.544786            5.0   
Braveheart                      3.0  0.654971  0.809303            4.0   
Dodgeball                       3.0  0.747619  0.864650            4.0   
Forest Gump                     4.0  0.954710  0.977093            5.0   
Gladiator                       2.0  0.666667  0.816497            5.0   
Jaws                            4.0  1.508772  1.228321            4.0   
Kazaam                          4.0  1.162500  1.078193            2.0   
Lord of the Rings               4.0  1.212121  1.100964            4.0   
Napolean Dynamite               4.0  1.780797  1.334465            4.0   
Old School                      3.0  0.941176  0.970143            5.0   
Pootie Tang                     4.0  1.944444  1.394433            2.0   
Pulp Fiction                    2.0  0.625000  0.790569            5.0   
Scarface                        3.0  0.606061  0.778499            4.0   
Shawshank Redemption            3.0  0.818182  0.904534            5.0   
Snakes on a Plane               3.0  1.123810  1.060099            3.0   
Spiderman                       3.0  0.782609  0.884652            4.0   
Star Wars                       2.0  0.521739  0.722315            5.0   
The Dark Knight                 2.0  0.368421  0.606977            5.0   
The Happening                   1.0  0.125000  0.353553            1.0   
The Matrix                      3.0  0.810277  0.900154            5.0   
Toy Story                       2.0  0.460000  0.678233            5.0   
Village                         3.0  1.266667  1.125463            3.5   
You Got Mail          

In [15]:
for movie in ['Shawshank Redemption', 'The Dark Knight', 'The Happening', 'Pootie Tang', 'Gladiator']:
    print ("----------------------------------------")
    print (movie)
    print (data.loc[movie].value_counts())
    print ("----------------------------------------")

----------------------------------------
Shawshank Redemption
5.0    8
4.0    3
2.0    1
Name: Shawshank Redemption, dtype: int64
----------------------------------------
----------------------------------------
The Dark Knight
5.0    9
4.0    9
3.0    1
Name: The Dark Knight, dtype: int64
----------------------------------------
----------------------------------------
The Happening
1.0    7
2.0    1
Name: The Happening, dtype: int64
----------------------------------------
----------------------------------------
Pootie Tang
1.0    6
5.0    1
3.0    1
2.0    1
Name: Pootie Tang, dtype: int64
----------------------------------------
----------------------------------------
Gladiator
4.0    7
3.0    6
5.0    6
Name: Gladiator, dtype: int64
----------------------------------------


### mean
* 2, 4, 4, 4, 5, 5, 5, 5, 5, 5, 5, 5
* sum of all = 54 / 12 
* mean or average = 4.5

### median
* 2, 4, 4, 4, 5, 5, 5, 5, 5, 5, 5, 5
* there are 12 (even) numbers, so let's take middle two numbers 6th, 7th and take the average
* median = 5 + 5 = 10 / 2 = 5

### mode
* 2, 4, 4, 4, 5, 5, 5, 5, 5, 5, 5, 5
* rating 5 has been given 8 times
* mode = 5

### range
* 2, 4, 4, 4, 5, 5, 5, 5, 5, 5, 5, 5
* highest = 5, lowest = 2
* range = 5 - 2 = 3

### quartile
* 2, 4, 4, 4, 5, 5, 5, 5, 5, 5, 5, 5
* 12 numbers, split it into four equal parts
* 2, 4, 4 | 4, 5, 5 | 5, 5, 5 | 5, 5, 5 |
* 4, 5, 5, 5

### skewness
* _
* _
* _